File: Aquarium/Engine/Sources/Engine/PourPoint.swift
import Foundation

struct PourPoint {
    let point: Point
    let waterFlow: [Point]
    let airFlow: [Point]

    init(point: Point, groups: [[Int]]) {
        self.point = point

        var waterFlow = [point]
        var airFlow = [point]

        let group = groups[point]

        let size = groups.count

        for row in 0..<size {
            for col in 0..<size {
                // skip points that are in different groups
                if groups[row][col] != group {
                    continue
                }

                // skip points that are the same as the first
                if row == point.row && col == point.col {
                    continue
                }

                if row >= point.row {
                    waterFlow.append(Point(row, col))
                }

                if row <= point.row {
                    airFlow.append(Point(row, col))
                }
            }
        }
        self.waterFlow = waterFlow
        self.airFlow = airFlow
    }

    func getPoints(_ state: State) -> [Point] {
        switch state {
        case .water: return waterFlow
        case .air: return airFlow
        case .none: return []
        }
    }
}

extension PourPoint: CustomDebugStringConvertible {
    var debugDescription: String {
        "PourPoint\(point) { water: \(waterFlow)}, air: \(airFlow)"
    }
}

File: rust/crates/grid/src/pour_point.rs
use crate::{Cell, Point};

#[derive(Debug)]
pub struct PourPoint {
    pub point: Point,
    pub water_flow: Vec<Point>,
    pub air_flow: Vec<Point>,
}

impl PourPoint {
    pub fn new(point: Point, cells: &Vec<Vec<Cell>>) -> Self {
        let size = cells.len();
        let group = cells[point.row][point.col].group;
        let (mut water_flow, mut air_flow) = (vec![], vec![]);

        for r in 0..size {
            for c in 0..size {
                if cells[r][c].group != group {
                    continue;
                }
                if r >= point.row {
                    water_flow.push(Point::new(r, c));
                }
                if r <= point.row {
                    air_flow.push(Point::new(r, c));
                }
            }
        }

        Self { point, water_flow, air_flow }
    }
}
